src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,341 @@
+                { "nvram-template", JSON_VARIANT_OBJECT, firmware_nvram_template, 0, JSON_MANDATORY },
+                {}
+        };
+
+        return json_dispatch(v, table, NULL, 0, userdata);
+}
+
+int find_ovmf_config(int search_sb, OvmfConfig** ret) {
+        _cleanup_(ovmf_config_freep) OvmfConfig* config = NULL;
+        int r;
+
+        config = new0(OvmfConfig, 1);
+        if (!config)
+                return -ENOMEM;
+
+        _cleanup_free_ char *xdg_config_dir = NULL, *user_firmware_dir;
+        r = xdg_user_config_dir(&xdg_config_dir, NULL);
+        if (r < 0)
+                return r;
+
+        user_firmware_dir = strjoin(xdg_config_dir, "/qemu/firmware");
+        if (!user_firmware_dir)
+                return -ENOMEM;
+
+        /* sdf_ptr is required to make CLEANUP_ARRAY work, without sdf_ptr the cleanup function is passed a
+         * ptr which has the bit representation of the first two elements of the array */
+        int source_dir_fds[3] = { -EBADF, -EBADF, -EBADF };
+        const int *sdf_ptr = source_dir_fds;
+        OvmfConfigFile* files = NULL;
+        size_t n_sdfs = ELEMENTSOF(source_dir_fds), n_cfs = 0, i = 0;
+        CLEANUP_ARRAY(sdf_ptr, n_sdfs, close_many);
+        CLEANUP_ARRAY(files, n_cfs, ovmf_config_file_free_many);
+
+        /* Search in:
+         * - $XDG_CONFIG_HOME/qemu/firmware
+         * - /etc/qemu/firmware
+         * - /usr/share/qemu/firmware
+         *
+         * Prioritising entries in "more specific" directories
+         */
+        char **search_dirs = STRV_MAKE(user_firmware_dir, "/etc/qemu/firmware", "/usr/share/qemu/firmware");
+        STRV_FOREACH(s, search_dirs) {
+                r = open(*s, O_RDONLY|O_DIRECTORY|O_CLOEXEC);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to open directory '%s', ignoring: %m", *s);
+                        i++;
+                        continue;
+                }
+                source_dir_fds[i++] = r;
+        }
+
+        for (i = 0; i < n_sdfs; i++) {
+                int sdf = source_dir_fds[i];
+                if (sdf == -EBADF)
+                        continue;
+
+                _cleanup_free_ DirectoryEntries* de = NULL;
+                r = readdir_all(sdf, RECURSE_DIR_SORT, &de);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to read the contents of '%s', ignoring: %m", search_dirs[i]);
+                        continue;
+                }
+
+                for (size_t j = 0; j < de->n_entries; j++) {
+                        const struct dirent *d = de->entries[j];
+                        if (!endswith(d->d_name, ".json"))
+                                continue;
+
+                        OvmfConfigFile *f = GREEDY_REALLOC(files, n_cfs + 1);
+                        if (!f)
+                                return -ENOMEM;
+
+                        f[n_cfs++] = (OvmfConfigFile) {
+                                .name = strdup(d->d_name),
```

Comment:
oom check

Reply:
Fixed in 69a7728f8a.

Reply:
Missed a free on the path this is fixed in c053f06da5.